x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -82,24 +81,28 @@
 import static java.util.Arrays.asList;
 import static java.util.Collections.singleton;
 import static org.elasticsearch.xpack.esql.expression.NamedExpressions.mergeOutputExpressions;
-import static org.elasticsearch.xpack.ql.common.Failure.fail;
 import static org.elasticsearch.xpack.ql.expression.Expressions.asAttributes;
-import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FOURTH;
-import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.THIRD;
-import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isFoldable;
 import static org.elasticsearch.xpack.ql.optimizer.OptimizerRules.FoldNull;
 import static org.elasticsearch.xpack.ql.optimizer.OptimizerRules.PropagateEquals;
 import static org.elasticsearch.xpack.ql.optimizer.OptimizerRules.PropagateNullable;
 import static org.elasticsearch.xpack.ql.optimizer.OptimizerRules.TransformDirection;
 
 public class LogicalPlanOptimizer extends ParameterizedRuleExecutor<LogicalPlan, LogicalOptimizerContext> {
 
+    private final LogicalVerifier verifier = LogicalVerifier.INSTANCE;
+
     public LogicalPlanOptimizer(LogicalOptimizerContext optimizerContext) {
         super(optimizerContext);
     }
 
     public LogicalPlan optimize(LogicalPlan verified) {
-        return verifyOptimized(verified.optimized() ? verified : execute(verified));
+        var optimized = execute(verified);
+
+        Collection<Failure> failures = verifier.verify(optimized);
```

Comment:
This runs the dependency check only after all optimization rules have run.

I think there's value in running the dependency check after each optimization rule, to be able pinpoint the offending optimization rule immediately - at least during tests, e.g. by using an assertion. What do you think?

This would probably require overriding `RuleExecutor.executeWithInfo`, so definitely out of scope for this PR.

Reply:
Not sure how useful that is considering there's a cost in setting and disabling the assertion - it will help localize the problem however as the code is stable I think it will be fairly easy to identify what happened without changing the rule executor.